cmd/clusterctl/pkg/client/cluster/inventory.go

```diff
@@ -0,0 +1,215 @@
+	List() ([]clusterctlv1.Provider, error)
+}
+
+// inventoryClient implements InventoryClient.
+type inventoryClient struct {
+	proxy Proxy
+}
+
+// ensure inventoryClient implements InventoryClient.
+var _ InventoryClient = &inventoryClient{}
+
+// newInventoryClient returns a inventoryClient.
+func newInventoryClient(proxy Proxy) *inventoryClient {
+	return &inventoryClient{
+		proxy: proxy,
+	}
+}
+
+func (p *inventoryClient) EnsureCRD() error {
+	// get the CRD manifest from the embedded assets
+	yaml, err := config.Asset("cmd/clusterctl/config/manifest/clusterctl-api.yaml")
+	if err != nil {
+		return err
+	}
+
+	// transform the yaml in a list of objects
+	objs, err := util.ToUnstructured(yaml)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse yaml for clusterctl inventory CRD")
+	}
+
+	c, err := p.proxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	for _, o := range objs {
+		klog.V(3).Infof("Creating: %s, %s/%s", o.GroupVersionKind(), o.GetNamespace(), o.GetName())
+		if err = c.Create(ctx, &o); err != nil { //nolint
+			if apierrors.IsAlreadyExists(err) {
+				return nil
+			}
+			return errors.Wrapf(err, "failed to create clusterctl inventory CRD")
+		}
+	}
+
+	return errors.Wrapf(err, "invalid yaml for clusterctl inventory CRD")
+}
+
+func (p *inventoryClient) Validate(m clusterctlv1.Provider) error {
+	instances, err := p.list(listOptions{
+		Name: m.Name,
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(instances) == 0 {
+		return nil
+	}
+
+	// Target Namespace check
+	// Installing two instances of the same provider in the same namespace won't be supported
+	for _, i := range instances {
+		if i.Namespace == m.Namespace {
+			return errors.Errorf("There is already an instance of the %q provider installed in the %q namespace", m.Name, m.Namespace)
+		}
+	}
+
+	// Version check:
+	// If we are going to install an instance of a provider with version X, and there are already other instances of the same provider with
+	// different versions there is the risk of creating problems to all the version different than X because we are going to override
+	// all the existing non-namespaced objects (e.g. CRDs) with the ones from version X
+	sameVersion := false
+	for _, i := range instances {
+		if i.Version == m.Version {
+			sameVersion = true
+		}
+	}
+	if !sameVersion {
+		return errors.Errorf("The new instance of the %q provider has a version different than other instances of the same provider", m.Name)
+	}
+
+	// Watching Namespace check:
+	// If we are going to install an instance of a provider watching objects in namespaces already controlled by other providers
+	// then there will be providers fighting for objects...
+	if m.WatchedNamespace == "" {
+		return errors.Errorf("The new instance of the %q provider is going to watch for objects in namespaces already controlled by other providers", m.Name)
+	}
+
+	sameNamespace := false
+	for _, i := range instances {
```

Comment:
No, because I'm checking for two instances of the same provider I'm validating fighting for reconciling the same objects (e.g two instances of the CAPA controllers fighting for objects). 
Having different providers watching the same namespace is ok

Reply:
ahh ok. thanks.